File: src/bindings/cursor.cpp
#include "../cursor.hpp"

#include "../document.hpp"

void Cursor::init_bridge(sol::table& core) {
    // clang-format off
    core.new_usertype<Cursor>("Cursor",
        /* Properties. */
        "row", sol::property([](const Cursor& self) -> std::size_t { return self.pos_.row_; }),
        "col", sol::property([](const Cursor& self) -> std::size_t { return self.pos_.col_; }),

        /* Functions. */
        "up", &Cursor::up,
        "down", &Cursor::down,
        "left", &Cursor::left,
        "right", &Cursor::right,

        "step_forward", &Cursor::step_forward,
        "step_backward", &Cursor::step_backward,
        "peek_forward", &Cursor::peek_forward,
        "peek_backward", &Cursor::peek_backward,
        "move_to", [](Cursor& self, const Document& doc, const std::size_t point) -> void { self.point(doc, point); },
        "point", [](const Cursor& self, const Document& doc) -> std::size_t { return self.point(doc); },

        "_jump_to_beginning_of_line", &Cursor::_jump_to_beginning_of_line,
        "_jump_to_end_of_line", &Cursor::_jump_to_end_of_line,
        "_jump_to_beginning_of_file", &Cursor::_jump_to_beginning_of_file,
        "_jump_to_end_of_file", &Cursor::_jump_to_end_of_file,
        "_next_word", &Cursor::_next_word,
        "_next_word_end", &Cursor::_next_word_end,
        "_prev_word", &Cursor::_prev_word,
        "_prev_word_end", &Cursor::_prev_word_end,
        "_next_whitespace", &Cursor::_next_whitespace,
        "_prev_whitespace", &Cursor::_prev_whitespace,
        "_next_empty_line", &Cursor::_next_empty_line,
        "_prev_empty_line", &Cursor::_prev_empty_line,
        "_jump_to_matching_opposite", &Cursor::_jump_to_matching_opposite);
    // clang-format on
}

File: lua/bindings/cursor.lua
--- @meta

--- A Cursor is the virtual index into a Document. It manages the byte offsets into the Document as well as a preferred
--- column to keep the Cursor in one column while scrolling.
--- @class Core.Cursor
--- @field row integer Physical row in the Document.
--- @field col integer Physical column in the Document.
Core.Cursor = {}

--- Moves the Cursor up one row.
--- @param view Core.DocumentView
--- @param n integer amount of rows
function Core.Cursor:up(view, n) end

--- Moves the Cursor down one row.
--- @param view Core.DocumentView
--- @param n integer amount of rows
function Core.Cursor:down(view, n) end

--- Moves the Cursor left by one utf8 character.
--- @param view Core.DocumentView
--- @param n integer amount of characters
function Core.Cursor:left(view, n) end

--- Moves the Cursor right by one utf8 character.
--- @param view Core.DocumentView
--- @param n integer amount of characters
function Core.Cursor:right(view, n) end

--- Moves the Cursor to the next utf8 character.
--- @param view Core.DocumentView
--- @return boolean success
function Core.Cursor:step_forward(view) end

--- Moves the Cursor to the previous utf8 character.
--- @param view Core.DocumentView
--- @return boolean success
function Core.Cursor:step_backward(view) end

--- Returns the utf8 code-point of the next utf8 character.
--- @param view Core.DocumentView
--- @return integer|nil code_point
function Core.Cursor:peek_forward(view) end

--- Returns the utf8 code-point of the previous utf8 character.
--- @param view Core.DocumentView
--- @return integer|nil code_point
function Core.Cursor:peek_backward(view) end

--- Moves the Cursor to specific point in the Document.
--- @param view Core.DocumentView
--- @param point integer
function Core.Cursor:move_to(view, point) end

--- Returns the point the Cursor in the Document is pointing at.
--- @param view Core.DocumentView
--- @return integer point
function Core.Cursor:point(view) end

-- Predefined movement functions for performance.

function Core.Cursor:_jump_to_beginning_of_line(view) end

function Core.Cursor:_jump_to_end_of_line(view) end

function Core.Cursor:_jump_to_beginning_of_file(view) end

function Core.Cursor:_jump_to_end_of_file(view) end

function Core.Cursor:_next_word(view, n) end

function Core.Cursor:_next_word_end(view, n) end

function Core.Cursor:_prev_word(view, n) end

function Core.Cursor:_prev_word_end(view, n) end

function Core.Cursor:_next_whitespace(view, n) end

function Core.Cursor:_prev_whitespace(view, n) end

function Core.Cursor:_next_empty_line(view, n) end

function Core.Cursor:_prev_empty_line(view, n) end

function Core.Cursor:_jump_to_matching_opposite(view) end
